mrmustard/physics/representations.py

```diff
@@ -373,27 +374,53 @@ def __getitem__(self, idx: int | tuple[int, ...]) -> Bargmann:
         new._contract_idxs = idx
         return new
 
-    def __matmul__(self, other: Bargmann) -> Bargmann:
+    def __matmul__(self, other: Union[Bargmann, Fock]) -> Union[Bargmann, Fock]:
         r"""
         The inner product of ansatze across the marked indices.
+
+        If ``other`` is ``Fock``, then ``self`` is converted to ``Fock`` before the contraction.
+
+        Args:
+            other: Another representation.
+
+        Returns:
+            A ``Bargmann`` representation if ``other`` is ``Bargmann``, or a ``Fock``representation
+            if ``other`` is ``Fock``.
         """
+        idx_s = self._contract_idxs
+        idx_o = other._contract_idxs
+
+        # if ``other`` is ``Fock``, convert ``self`` to ``Fock``
+        if isinstance(other, Fock):
+            from .converters import to_fock  # pylint: disable=import-outside-toplevel
+
+            # set same shape along the contracted axes, and default shape along the
+            # axes that are not being contracted
+            shape = [settings.AUTOCUTOFF_MAX_CUTOFF for _ in range(len(self.b[0]))]
+            for i, j in zip(idx_s, idx_o):
+                shape[i] = other.array.shape[1:][j]
+
+            return to_fock(self, shape=shape)[idx_s] @ other[idx_o]
+
         if self.ansatz.degree > 0 or other.ansatz.degree > 0:
```

Comment:
This is awkward, because technically the Fock representation is all polynomial..